dbr quant: support dynamic linear

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70372 dbr quant: add limited support for `torch.nn.ModuleList`
* #70349 dbr quant support for custom leaf modules, part 3/x
* #70335 dbr quant: support for custom leaf modules, part 2/x
* #70330 dbr quant: support for custom leaf modules, part 1/x
* #70329 dbr quant: unbreak case when child module not returning any outputs
* #70328 dbr quant: support functional linear without bias
* **#70257 dbr quant: support dynamic linear**
* #70256 dbr quant: do not attach auto_quant_state to observers

Summary:

Makes dynamic quantization for linear module work in DBR quant.

Coverage for more ops and functionals will be in future PRs.

Test plan:

```
python test/test_quantization.py -k DBR
```

Differential Revision: [D33262300](https://our.internmc.facebook.com/intern/diff/D33262300)

vkuzo committed:
dbr quant: support dynamic linear

Summary:

Makes dynamic quantization for linear module work in DBR quant.

Coverage for more ops and functionals will be in future PRs.

Test plan:

```
python test/test_quantization.py -k DBR
```

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33262300).

vkuzo committed:
Update on "dbr quant: support dynamic linear"


Summary:

Makes dynamic quantization for linear module work in DBR quant.

Coverage for more ops and functionals will be in future PRs.

Test plan:

```
python test/test_quantization.py -k DBR
```

Differential Revision: [D33262300](https://our.internmc.facebook.com/intern/diff/D33262300)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33262300).

vkuzo committed:
Update on "dbr quant: support dynamic linear"


Summary:

Makes dynamic quantization for linear module work in DBR quant.

Coverage for more ops and functionals will be in future PRs.

Test plan:

```
python test/test_quantization.py -k DBR
```

Differential Revision: [D33262300](https://our.internmc.facebook.com/intern/diff/D33262300)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33262300).

vkuzo committed:
Update on "dbr quant: support dynamic linear"


Summary:

Makes dynamic quantization for linear module work in DBR quant.

Coverage for more ops and functionals will be in future PRs.

Test plan:

```
python test/test_quantization.py -k DBR
```

Differential Revision: [D33262300](https://our.internmc.facebook.com/intern/diff/D33262300)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33262300).

jerryzh168 commented on torch/ao/quantization/_dbr/model_utils.py:
```diff
@@ -102,10 +102,13 @@ def attach_scale_zp_values_to_model(
-            # tensor_id_to_observer is a ModuleDict which has to have string keys
-            # tensor_id_to_scale_zp is a normal dict which can have int keys
-            qstate.tensor_id_to_scale_zp[int(tensor_id)] = (scale, zp)
+            activation_int8_quantized = \
```
I think we also have some utility functions for this (may need some cleanup): https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/utils.py#L98

vkuzo replied:
yeah, existing function takes qconfig, not sure if it's worth creating a new util function with only one callsite for now

jerryzh168 replied:
I see, sounds good, is there a way to keep track of this so that we don't miss it in the future when we expand this condition?

vkuzo replied:
maybe we can think about it when there is a second callsite? Not sure if it's worth to do this now when we don't when/if the second callsite would appear.

vkuzo replied:
(motivation: https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)

jerryzh168 replied:
sure, sounds good

jerryzh168 commented on torch/ao/quantization/_dbr/module_swap_utils.py:
```diff
@@ -23,16 +27,21 @@ def _swap_child_modules(
         if not qconfig:
             continue
         activation_int8_quantized = activation_is_int8_quantized(qconfig)
+        op_int8_dynamically_quantized = op_is_int8_dynamically_quantized(qconfig)
```
might be good to move op_is_int8_dynamically_quantized function to https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/utils.py so that other tools can use it as well

jerryzh168 commented on torch/ao/quantization/utils.py:
```diff
@@ -113,6 +113,15 @@ def weight_is_statically_quantized(qconfig):
+def op_is_int8_dynamically_quantized(qconfig) -> bool:
+    activation_dtype, weight_dtype, activation_compute_dtype = \
+        get_qconfig_dtypes(qconfig)
+    return (
+        activation_dtype is torch.float and
+        weight_dtype in (torch.quint8, torch.qint8) and
+        activation_compute_dtype in (torch.quint8, torch.qint8)
+    )
```
assuming this is for fbgemm/qnnpack, should we explicitly list all possible combinations?
we only support activation_compute_dtype == torch.quint8 and weight_dtype == torch.qint8? or do we also support other combinations like (activation_compute_dtype == torch.qint8 and weight_dtype == torch.quint8)?

vkuzo replied:
sure, for now this is for fbgemm and qnnpack so I can update the code to reflect that, we can make it more general in a future PR

jerryzh168 reviewed (approved):
had a few comments inline

vkuzo committed:
Update on "dbr quant: support dynamic linear"


Summary:

Makes dynamic quantization for linear module work in DBR quant.

Coverage for more ops and functionals will be in future PRs.

Test plan:

```
python test/test_quantization.py -k DBR
```

Differential Revision: [D33262300](https://our.internmc.facebook.com/intern/diff/D33262300)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33262300).
